### PR TITLE
Modified pull experiment to get also cell sets

### PR DESCRIPTION
**Pull request by kafkasl**

### CODE REVIEW

**Review by cosa65 — COMMENTED** [review]
I am getting some issues related to the worker files (not mock_cell_sets) but otherwise it seems to be working correctly on my end.

Apart from this, it seems that the `mock_cell_sets.json` file is actually saved as a binary so we should change the extension maybe (or change it so that it is stored as a json and change the inframock PR which has already been changed to account for the file being a binary).

**Comment by kafkasl** [comment]
@cosa65 cell sets are now saved in plain text. I fixed other minor issues like detecting whether a dynamoDB file needs updating (until now they were always updated)

File: biomage/experiment/pull.py [hunk 1]
```diff
@@ -101,8 +104,13 @@ def pull(experiment_id, origin):
     Summary.set_command(cmd=PULL, origin=origin, experiment_id=experiment_id)
     bucket = f"biomage-source-{origin}"
     file = f"{experiment_id}/r.rds"
+    dst_file = file + ".gz"
+    # download_if_modified(bucket=bucket, key=file, filepath=dst_file)
```
[the remark this answers is not in the document]
Comment (kafkasl) [reply biomage/experiment/pull.py]: actually that's the line that downloads the dataset itself. I forgot to uncomment it

File: biomage/experiment/pull.py [hunk 2]
```diff
@@ -52,16 +73,16 @@ def update_config_if_needed(file, table_name, experiment_id):
         remote_cfg = remove_key(remote_cfg, "pipeline")
 
     # if the local config was not found or it's different from the remote => update
-    if not found or DeepDiff(local_cfg, remote_cfg):
-        save_cfg_file(remote_cfg, file)
-        Summary.add_changed_file(file)
+    if not found or local_cfg != remote_cfg:
```
Comment (ivababukova) [inline biomage/experiment/pull.py:76]: why is there no need for this DeepDiff anymore?

File: biomage/experiment/pull.py [hunk 3]
```diff
@@ -52,16 +73,16 @@ def update_config_if_needed(file, table_name, experiment_id):
...
+        Summary.add_changed_file(filepath)
 
 
 def update_configs(experiment_id, origin):
     # config pairs like: (local file name, remote table name)
     configs = [
-        ("mock_experiment.json", "experiments"),
-        ("mock_samples.json", "samples"),
+        (EXPERIMENTS_FILE, EXPERIMENTS_TABLE),
+        (SAMPLES_FILE, SAMPLES_TABLE),
```
Comment (ivababukova) [inline biomage/experiment/pull.py:85]: that's much better than hardcoding them. Eventually, we can move them to config file (I know that Agi created one as part of his PR)

File: biomage/experiment/pull.py [hunk 4]
```diff
@@ -99,10 +120,17 @@ def pull(experiment_id, origin):
     """
 
     Summary.set_command(cmd=PULL, origin=origin, experiment_id=experiment_id)
+
     bucket = f"biomage-source-{origin}"
-    file = f"{experiment_id}/r.rds"
+    file = f"{experiment_id}/{RDS_FILE}"
+    dst_file = file + ".gz"
+    download_if_modified(bucket=bucket, key=file, filepath=dst_file)
```
Comment (ivababukova) [inline biomage/experiment/pull.py:127]: the name of the bucket here and in line 133 are different, even though we put in the function 
 calls of `download_if_modified` `bucket=bucket` in both places. What this means that if we swapped the two lines of code, technically the code will run correctly, but the results will be wrong in a subtle way. I can think of these ways in mitigating this:
 - directly put the bucket name in each of the calls to `download_if_modified` and not have a `bucket` variable defined
 - extract out the two different bucket names as different variables and put them at the top of the file similarly to what is done for `PLOTS_TABLES_FILE`, `CELLSETS_FILE`, etc. for example. Then in `download_if_modified` refer to the relevant bucket name.
 - less effective, but you can add a comment above each call to `download_if_modified` 
What do you think?

**Review by cosa65 — APPROVED** [review]

File: biomage/experiment/pull.py (hunk 3 above)
Comment (kafkasl) [reply biomage/experiment/pull.py:85]: yes, that would be optimal

File: biomage/experiment/pull.py (hunk 2 above)
Comment (kafkasl) [reply biomage/experiment/pull.py:76]: DeepDiff was not considering equal fields with same value but different type (educated guess) considering different `Decimal(10)` and just `10` and thus failing the comparison. However, just using equal works fine (the structures are saved / loaded the same way so it seems they can indeed be compared like that).

Comment (ivababukova) [reply biomage/experiment/pull.py:76]: you can also remove it from requirements.txt in this case too: https://github.com/biomage-ltd/biomage-utils/blob/master/requirements.txt#L16

**Pull request merged**